### PR TITLE
Mdx stanza

### PR DESCRIPTION
**Pull request by NathanReb**
This is still draft as I need to either mock mdx for the tests or disable them by default.

I also need to get some stuff merged on the mdx side and cut a release before we can merge this.

Finally some mdx users call dune within mdx files and I want to make sure this won't break too much stuff either.

### CODE REVIEW

**Comment by NathanReb** [comment]
Note to myself: I need to bump the dune lang as well now that 2.2.0 has been released.

**Comment by aalekseyev** [comment]
Is having mdx enabled in CI problematic for some reason? (is it particularly big or broken?)
If adding it to opam_install_test_deps .travis-ci.sh is sufficient to make tests pass, I think that's probably the way to go instead of disabling tests.

**Comment by NathanReb** [comment]
> Is having mdx enabled in CI problematic for some reason?
It shouldn't be, you're right. I need to release the version I used to work on the stanza but then I'll add mdx to the CI.

File: src/dune/mdx.ml [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    | [ line ] ->
+      let items = String.split ~on:' ' line in
+      Result.List.map ~f:parse_one items
+    | _ ->
+      Result.errorf
+        "Invalid 'ocaml-mdx deps' output should a single line with space \
+         separated items"
+
+  let read files =
+    let open Files in
```
Comment (rgrinberg) [inline src/dune/mdx.ml]: Let's get rid of this open and just annotate the type of `files : Files.t`

File: src/dune/mdx.ml [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+  List.filter_map src_dir_files ~f:(fun src_path ->
+      if must_mdx src_path then
+        Some (build_path src_path)
+      else
+        None)
+
+(** Generates the rules for a single [src] file covered covered by the given
+    [stanza]. *)
+let gen_rules_for_single_file ~sctx ~dir ~mdx_prog ~stanza src =
+  let open Files in
```
Comment (rgrinberg) [inline src/dune/mdx.ml]: Same as above. The open is confusing

**Review by rgrinberg — COMMENTED** [review]
Looks good. Just some minor changes required and let's make sure the manual is up to date.

File: src/dune/mdx.ml [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    Path.build build
+
+  let dirs_and_files ~dir t_list =
+    List.partition_map t_list ~f:(function
+      | Dir d -> Left (to_path ~dir d)
+      | File f -> Right (to_path ~dir f))
+
+  let dir_without_files_dep =
+    let pred =
+      Predicate.create ~id:(lazy (String "false")) ~f:(fun _ -> false)
```
[the remark this answers is not in the document]
Comment (NathanReb) [reply src/dune/mdx.ml]: Good idea!

File: src/dune/mdx.ml [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+let decode =
+  let open Dune_lang.Decoder in
+  fields
+    (let+ loc = loc
+     and+ files =
+       field "files" Predicate_lang.Glob.decode ~default:default_files
+     and+ packages = field_o "packages" (repeat Package.Name.decode)
+     and+ preludes = field_o "preludes" (repeat Prelude.decode) in
+     let packages = Option.value ~default:[] packages in
```
Comment (rgrinberg) [inline src/dune/mdx.ml]: I believe you can use `field ~default:[]` to simplify the definition of packages and preludes.

File: src/dune/mdx.ml [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            ; Target files.corrected
+            ; Dep (Path.build files.src)
+            ] )
+  in
+  Super_context.add_rule sctx ~loc ~dir mdx_action;
+  (* Attach the diff action to the @runtest for the src and corrected files *)
+  let diff_action =
+    let+ () = Build.path (Path.build files.src)
+    and+ () = Build.path (Path.build files.corrected) in
+    Action.diff ~optional:false (Path.build files.src)
```
Comment (rgrinberg) [inline src/dune/mdx.ml]: you might as well do:

```
let src = Path.build files.src
and corrected = Path.build files.corrected
in
```

**Comment by NathanReb** [comment]
I just changed the communication format to use canonical sexp instead of something adhoc and fragile.

If that seems fine by you I'll cut an MDX release so that I can fix the CI here and we can finally merge!

**Comment by NathanReb** [comment]
Still have to write the documentation though!

**Comment by NathanReb** [comment]
I added some documentation, removed mli files from the doc and the default files as the mli support didn't make it to `mdx.1.6.0`.

Last step is for opam-repo to properly update so that we can install mdx.1.6.0 in the CI to get the test to pass there!

**Comment by NathanReb** [comment]
I'm undrafting as it is essentially ready for review!

File: src/dune/mdx.ml [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    Path.build build
+
+  let dirs_and_files ~dir t_list =
+    List.partition_map t_list ~f:(function
+      | Dir d -> Left (to_path ~dir d)
+      | File f -> Right (to_path ~dir f))
+
+  let dir_without_files_dep =
+    let pred =
+      Predicate.create ~id:(lazy (String "false")) ~f:(fun _ -> false)
```
Comment (rgrinberg) [inline src/dune/mdx.ml]: This should be moved to `Predicate.false_` (like `Predicate.true_)

**Review by rgrinberg — APPROVED** [review]

**Comment by NathanReb** [comment]
`mdx.1.6.0` has been released so the CI should go green now!

**Comment by NathanReb** [comment]
Just learned about the `runtest-no-deps` thing. Now it should be good!

**Comment by NathanReb** [comment]
And there we go! @rgrinberg can I go ahead and merge?

**Comment by rgrinberg** [comment]
Yup

**Pull request merged**